lib/AST/ASTImporter.cpp

```diff
@@ -2180,7 +2180,29 @@ Decl *ASTNodeImporter::VisitRecordDecl(RecordDecl *D) {
         if (!ToDescribed)
           return nullptr;
         D2CXX->setDescribedClassTemplate(ToDescribed);
-
+        if (!DCXX->isInjectedClassName() && D->isCompleteDefinition()) {
+          // In a record describing a template the type should be a
+          // InjectedClassNameType (see Sema::CheckClassTemplate). Update the
+          // previously set type to the correct value here (ToDescribed is not
+          // available at record create).
+          // FIXME: The previous type is cleared but not removed from
+          // ASTContext's internal storage.
+          D2CXX->setTypeForDecl(nullptr);
+          Importer.getToContext().getInjectedClassNameType(D2CXX,
+              ToDescribed->getInjectedClassNameSpecialization());
+          // Find the injected class name itself and update the type too.
+          CXXRecordDecl *Injected = [&] {
+            for (NamedDecl *Found : D2CXX->noload_lookup(Name)) {
+              auto *Record = dyn_cast<CXXRecordDecl>(Found);
+              if (Record && Record->isInjectedClassName()) {
+                return Record;
+              }
+            }
+            assert(false && "No injected-class-name found!");
+          }();
+          Injected->setTypeForDecl(nullptr);
```

Comment:
The DeclContext of Injected type should be the original (parent) class right? So is it possible that we can access the parent class and set the type for decl properly when the injected class created?

Reply:
This is done already now, line 2163 does a copy of the type from parent class. The type for the parent class is still not `InjectedClassNameType` at this place. The type of the injected-record should be the same as the type of its parent, the `Importer.getToContext().getTypeDeclType(Injected, D2CXX)` call copies the type into the injected record.